packages/vscode/src/index.ts

```diff
@@ -1,71 +1,103 @@
+			},
+			dontFilterIncompleteCompletions: true, // VSCode filters client side and is smarter at it than us
+		},
+	};
+
+	let client = createLanguageServer(serverOptions, clientOptions);
+	context.subscriptions.push(client.start());
+
+	client
+		.onReady()
+		.then(() => {
+			const tagRequestor = (document: TextDocument, position: Position) => {
+				const param = client.code2ProtocolConverter.asTextDocumentPositionParams(document, position);
+				return client.sendRequest(TagCloseRequest, param);
+			};
+			const disposable = activateTagClosing(tagRequestor, { astro: true }, 'html.autoClosingTags');
+			context.subscriptions.push(disposable);
+		})
+		.catch((err) => {
+			console.error('Astro, unable to load language server.', err);
+		});
+
+	// Restart the language server if any critical files that are outside our jurisdiction got changed (tsconfig, jsconfig etc)
+	workspace.onDidSaveTextDocument(async (doc: TextDocument) => {
+		const fileName = doc.fileName.split('/').pop() ?? doc.fileName;
+		if ([/^tsconfig\.json$/, /^jsconfig\.json$/, /^astro\.config\.(js|cjs|mjs|ts)$/].some((regex) => regex.test(fileName))) {
+			await restartClient(false);
+		}
+	});
+
+	context.subscriptions.push(
+		commands.registerCommand('astro.restartLanguageServer', async () => {
+			await restartClient(true);
+		})
+	);
+
+	let restartingClient = false;
+	async function restartClient(showNotification: boolean) {
+		if (restartingClient) {
+			return;
+		}
+
+		restartingClient = true;
+		await client.stop();
+
+		client = createLanguageServer(serverOptions, clientOptions);
+		context.subscriptions.push(client.start());
+		await client.onReady();
+
+		if (showNotification) {
+			window.showInformationMessage('Astro language server restarted.');
+		}
+
+		restartingClient = false;
+	}
+
+	function getLSClient() {
+		return client;
+	}
+
+	return {
+		getLanguageServer: getLSClient,
+	};
 }
 
-/**  */
-function createLanguageService(context: vscode.ExtensionContext, mode: 'doc', id: string, name: string, port: number) {
-  const { workspace } = vscode;
-  const serverModule = require.resolve('@astrojs/language-server/bin/server.js');
-  const debugOptions = { execArgv: ['--nolazy', '--inspect=' + port] };
-  const serverOptions: lsp.ServerOptions = {
-    run: { module: serverModule, transport: lsp.TransportKind.ipc },
-    debug: {
-      module: serverModule,
-      transport: lsp.TransportKind.ipc,
-      options: debugOptions,
-    },
-  };
-  const serverInitOptions: any = {
-    mode: mode,
-    appRoot: vscode.env.appRoot,
-    language: vscode.env.language,
-  };
-  const clientOptions: lsp.LanguageClientOptions = {
-    documentSelector: [{ scheme: 'file', language: 'astro' }],
-    synchronize: {
-      configurationSection: ['astro', 'javascript', 'typescript', 'prettier'],
-      fileEvents: workspace.createFileSystemWatcher('{**/*.js,**/*.ts}', false, false, false),
-    },
-    initializationOptions: {
-      ...serverInitOptions,
```

Comment:
Removed `serverInitOptions` from the final result since.. Well, I'm not sure what it does? Those variables are not used anywhere else and additionally, they're not in the VS Code's types for `initializationOptions`, not sure where they came from